### PR TITLE
fix: scale types for pending ceremonies

### PR DESCRIPTION
**Pull request by martin-chainflip**
# Pull Request

Closes: PRO-1676

## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [ ] I have written sufficient tests.
- [ ] I have written and tested required migrations.
- [ ] I have updated documentation where appropriate.

## Summary

This change renames the types returned by the ThresholdSignature -> pendingCeremonies call so that different types have different names.

### CODE REVIEW

File: state-chain/pallets/cf-threshold-signature/src/lib.rs [hunk 1]
```diff
@@ -364,6 +363,8 @@ pub mod pallet {
 		/// A marker trait identifying the chain that we are signing for.
 		type TargetChainCrypto: ChainCrypto;
 
+		type TargetChain: Chain;
+
```
Comment (dandanlen) [inline state-chain/pallets/cf-threshold-signature/src/lib.rs:367]:
```suggestion
		/// A marker trait identifying the chain that we are signing for.
		type TargetChainCrypto: ChainCrypto;

		type TargetChain: Chain<ChainCrypto = Self::TargetChainCrypto>;

```

Just to make sure can't do something stupid...

File: state-chain/runtime/src/lib.rs [hunk 2]
```diff
@@ -788,6 +788,7 @@ impl pallet_cf_threshold_signature::Config<Instance16> for Runtime {
 	type ThresholdCallable = RuntimeCall;
 	type ThresholdSignerNomination = chainflip::RandomSignerNomination;
 	type TargetChainCrypto = EvmCrypto;
+	type TargetChain = Arbitrum;
```
Comment (dandanlen) [inline state-chain/runtime/src/lib.rs:791]: What about Ehthereum?

File: state-chain/pallets/cf-threshold-signature/src/lib.rs (hunk 1 above)
Comment (dandanlen) [reply state-chain/pallets/cf-threshold-signature/src/lib.rs:367]: Actually I'm not sure we need (or *can* provide) the target chain here. Multiple chains can use the same instance of this pallet.

File: state-chain/pallets/cf-threshold-signature/src/lib.rs [hunk 3]
```diff
@@ -1199,6 +1200,66 @@ pub mod pallet {
...
+macro_rules! append_chain_to_name {
+	($name:ident) => {
+		match T::TargetChain::NAME {
+			"Ethereum" => concat!(stringify!($name), "Ethereum"),
+			"Polkadot" => concat!(stringify!($name), "Polkadot"),
+			"Bitcoin" => concat!(stringify!($name), "Bitcoin"),
+			"Arbitrum" => concat!(stringify!($name), "Arbitrum"),
+			"Solana" => concat!(stringify!($name), "Solana"),
+			_ => concat!(stringify!($name), "Other"),
+		}
```
Comment (dandanlen) [inline state-chain/pallets/cf-threshold-signature/src/lib.rs:1212]:
```suggestion
		concat!(stringify!($name), T::TargetChain::NAME)
```

But as mentioned elsewhere it will have to be ChainCrypto not Chain.

File: state-chain/pallets/cf-threshold-signature/src/lib.rs (hunk 1 above)
Comment (martin-chainflip) [reply state-chain/pallets/cf-threshold-signature/src/lib.rs:367]: that's fine, though. If it is the same pallet, then it will also have the same types and the same serialization. All we need is for different instances of this generic pallet to have different names. The most straightforward way would be to use the NAME parameter of the 'Chain' type (ChainCrypto does not provide a name)

File: state-chain/runtime/src/lib.rs (hunk 2 above)
Comment (martin-chainflip) [reply state-chain/runtime/src/lib.rs:791]: We don't instantiate `pallet_cf_threshold_signature::Config` for `Instance1` anywhere.

Comment (dandanlen) [reply state-chain/runtime/src/lib.rs:791]: Exactly: it's instantiated for the EVM, it's neither Arbitrum nor Ethereum.

File: state-chain/pallets/cf-threshold-signature/src/lib.rs (hunk 1 above)
Comment (dandanlen) [reply state-chain/pallets/cf-threshold-signature/src/lib.rs:367]: Then we should add a name? Otherwise it's just confusing. We will see 'Arbitrum' even when it's Ethereum. If we're going to add names, we might as well make them accurate?

**Review by dandanlen — COMMENTED** [review]
TLDR - I think we should add a name to the ChainCrypto trait instead of just using Arbitrum.

**Comment by dandanlen** [comment]
In fact it doesn't even have to be a NAME string, it just has to be something we can match on.

File: state-chain/pallets/cf-threshold-signature/src/lib.rs [hunk 4]
```diff
@@ -1199,6 +1197,65 @@ pub mod pallet {
...
+{
+	type Identity = Self;
+	fn type_info() -> Type {
+		Type::builder()
+			.path(Path::new(append_chain_to_name!(RequestContext), module_path!()))
+			.composite(
+				Fields::named()
+					.field(|f| f.ty::<RequestId>().name("request_id"))
+					.field(|f| f.ty::<AttemptCount>().name("attempt_count"))
+					.field(|f| f.ty::<PayloadFor<T, I>>().name(append_chain_to_name!(PayloadFor))),
```
Comment (dandanlen) [inline state-chain/pallets/cf-threshold-signature/src/lib.rs:1254]: Sorry, another comment... 

I don't understand why we need to use this `append_chain_to_name` on the field names... surely it's only needed on the path? 

If we want to resolve the type `PayloadFor<T, I>` then we need to apply the same treatment to the `PayloadFor<T, I>` *types* and somehow direct the type_info to resolve that type.

Comment (martin-chainflip) [reply state-chain/pallets/cf-threshold-signature/src/lib.rs:1254]: The only way to direct type_info on how to resolve a type is via its name. If we don't append the chain name to PayloadFor, then there will be multiple types called "PayloadFor" and the recursive type resolution will fail again.

Comment (martin-chainflip) [reply state-chain/pallets/cf-threshold-signature/src/lib.rs:1254]: my previous message was not correct. I added separate "field names" and "type names" now

**Review by dandanlen — APPROVED** [review]
